Mathlib/RingTheory/FiniteLength.lean

```diff
@@ -76,3 +87,85 @@ theorem isFiniteLength_iff_exists_compositionSeries :
+        sSup s = ⊤ ∧ ∀ m ∈ s, IsSimpleModule R m] := by
+  rw [isFiniteLength_iff_isNoetherian_isArtinian]
+  obtain ⟨s, hs⟩ := IsSemisimpleModule.exists_setIndependent_sSup_simples_eq_top R M
+  tfae_have 1 ↔ 2 := ⟨fun _ ↦ inferInstance, fun _ ↦ inferInstance⟩
+  tfae_have 2 → 5 := fun _ ↦ ⟨s, CompleteLattice.WellFoundedGT.finite_of_setIndependent hs.1, hs⟩
+  tfae_have 3 → 5 := fun _ ↦ ⟨s, CompleteLattice.WellFoundedLT.finite_of_setIndependent hs.1, hs⟩
+  tfae_have 5 → 4 := fun ⟨s, fin, _, sSup_eq_top, simple⟩ ↦ by
+    rw [← isNoetherian_top_iff, ← Submodule.topEquiv.isArtinian_iff,
+      ← sSup_eq_top, sSup_eq_iSup, ← iSup_subtype'']
+    rw [SetCoe.forall'] at simple
+    have := fin.to_subtype
+    exact ⟨isNoetherian_iSup, isArtinian_iSup⟩
+  tfae_have 4 → 2 := And.left
+  tfae_have 4 → 3 := And.right
+  tfae_finish
+
+instance [IsSemisimpleModule R M] [Module.Finite R M] : IsArtinian R M :=
+  (IsSemisimpleModule.finite_tfae.out 0 2).mp ‹_›
+
+example [IsSemisimpleRing R] : IsNoetherianRing R := inferInstance
+example [IsSemisimpleRing R] : IsArtinianRing R := inferInstance
+
+open Ideal in
+theorem IsSemiprimaryRing.isNoetherian_iff_isArtinian [IsSemiprimaryRing R] :
+    IsNoetherian R M ↔ IsArtinian R M := by
+  have ⟨ss, n, hn⟩ := (isSemiprimaryRing_iff R).mp ‹_›
+  set Jac := Ring.jacobson R
+  replace hn : Jac ^ n ≤ Module.annihilator R M := hn ▸ bot_le
+  induction' n with n ih generalizing M
+  · rw [Submodule.pow_zero, one_eq_top, top_le_iff, Module.annihilator_eq_top_iff] at hn
+    constructor <;> infer_instance
+  obtain _ | n := n
+  · rw [Submodule.pow_one, ← SetLike.coe_subset_coe,
+      ← Module.isTorsionBySet_iff_subset_annihilator] at hn
+    let _ := hn.module
+    have := hn.isSemisimpleModule_iff.mp inferInstance
+    exact IsSemisimpleModule.finite_tfae.out 1 2
+  let N := Jac ^ (n + 1) • (⊤ : Submodule R M)
+  simp_rw [iff_iff_eq] at ih -- otherwise `rw [ih]` below fails!
+  rw [isNoetherian_iff_submodule_quotient N, isArtinian_iff_submodule_quotient N, ih, ih]
+  · rw [← SetLike.coe_subset_coe, ← Module.isTorsionBySet_iff_subset_annihilator,
+      Module.isTorsionBySet_quotient_iff]
+    intro m i hi; exact Submodule.smul_mem_smul hi trivial
+  · rw [← Submodule.annihilator_top, Submodule.le_annihilator_iff, Ideal.IsTwoSided.pow_succ,
+      Submodule.mul_smul, ← Submodule.le_annihilator_iff] at hn
+    exact (Ideal.pow_le_self n.succ_ne_zero).trans hn
+
+variable (R M)
+
+theorem IsArtinianRing.tfae [IsArtinianRing R] :
+    List.TFAE [Module.Finite R M, IsNoetherian R M, IsArtinian R M, IsFiniteLength R M] := by
+  tfae_have 2 ↔ 3 := IsSemiprimaryRing.isNoetherian_iff_isArtinian
+  tfae_have 2 → 1 := fun _ ↦ inferInstance
+  tfae_have 1 → 3 := fun _ ↦ inferInstance
+  rw [isFiniteLength_iff_isNoetherian_isArtinian]
+  tfae_have 4 → 2 := And.left
+  tfae_have 2 → 4 := fun h ↦ ⟨h, tfae_2_iff_3.mp h⟩
+  tfae_finish
+
+instance [IsArtinianRing R] : IsNoetherianRing R := ((IsArtinianRing.tfae R R).out 2 1).mp ‹_›
```

Comment:
Also cc @jjaassoonn because I think you proved the commutative version (due to Akizuki).